samcli/commands/local/lib/local_lambda.py

```diff
@@ -234,42 +238,48 @@ def _make_env_vars(self, function: Function) -> EnvironmentVariables:
         """
 
         function_id = function.function_id
-        name = function.name
+        logical_id = function.name
+        function_name = function.functionname
         full_path = function.full_path
 
         variables = None
         if isinstance(function.environment, dict) and "Variables" in function.environment:
             variables = function.environment["Variables"]
         else:
-            LOG.debug("No environment variables found for function '%s'", name)
+            LOG.debug("No environment variables found for function '%s'", logical_id)
 
-        # This could either be in standard format, or a CloudFormation parameter file format.
+        # This could either be in standard format, or a CloudFormation parameter file format, or mix of both.
         #
         # Standard format is {FunctionName: {key:value}, FunctionName: {key:value}}
         # CloudFormation parameter file is {"Parameters": {key:value}}
+        # Mixed format is {FunctionName: {key:value}, "Parameters": {key:value}}
 
         for env_var_value in self.env_vars_values.values():
             if not isinstance(env_var_value, dict):
-                reason = """
-                            Environment variables must be in either CloudFormation parameter file
-                            format or in {FunctionName: {key:value}} JSON pairs
-                            """
+                reason = "Environment variables {} in incorrect format".format(env_var_value)
                 LOG.debug(reason)
                 raise OverridesNotWellDefinedError(reason)
 
+        overrides = {}
+        # environment variables for specific resources take precedence over
+        # the single environment variable for all resources
         if "Parameters" in self.env_vars_values:
-            LOG.debug("Environment variables overrides data is in CloudFormation parameter file format")
+            LOG.debug("Environment variables data found in the CloudFormation parameter file format")
             # CloudFormation parameter file format
-            overrides = self.env_vars_values["Parameters"]
-        else:
+            parameter_result = self.env_vars_values.get("Parameters", {})
+            overrides.update(parameter_result)
+
+        # Precedence: logical_id -> function_id -> function name -> full_path, customer can use any of them
+        fn_file_env_vars = (
+            self.env_vars_values.get(logical_id, None)
+            or self.env_vars_values.get(function_id, None)
+            or self.env_vars_values.get(function_name, None)
```

Comment:
we need to update our docs to let the customers know that they can use function name in the env vars file.

Reply:
I will add this to the doc update ticket. Thanks @moelasmar